### PR TITLE
Pre preamble

### PR DESCRIPTION
**Pull request by jerryduan07**
implemented pre-preamble in packet assembler
content of pre-preamble depends on the first bit of packet: 11110000 if first bit 0, 00001111 if first bit 1
the inclusion of pre-preamble increases the robustness of ble baseband, provides an extra edge for cdr modules, and avoid potential scenarios where the first bit of packet preamble is missed and the entire packet is compromised

### CODE REVIEW

**Review by edwardcwang — CHANGES_REQUESTED** [review]
Packet disassembler is CDR output from RF -> DMA right? If so, since pre-preamble is NOT part of the spec (IIRC), I'd be very hesistant to include this. I think for the transmit part we can be free to transmit the pre-preamble but requiring it in RX would be a violation of the spec. If you insist, make it an optional parameter you can toggle in the CPU.

**Comment by jerryduan07** [comment]
> Packet disassembler is CDR output from RF -> DMA right? If so, since pre-preamble is NOT part of the spec (IIRC), I'd be very hesistant to include this. I think for the transmit part we can be free to transmit the pre-preamble but requiring it in RX would be a violation of the spec. If you insist, make it an optional parameter you can toggle in the CPU.

It's not required in RX and will not influence the packet disassembler. The pattern of pre-preamble will not be recognized by packet disassembler as a valid preamble, so the packet disassembler still starts at regular preamble, and this additional byte of data will provide some robustness and insurance

**Comment by Fibonacci12358** [comment]
@jerryduan07 I think you mean that this changes the assembler right? Looking at the code changes, I think the above mention of the disassembler was a typo.
I would remove the prepreamble from the disassembler tester though.

**Review by edwardcwang — DISMISSED** [review]
I see. In that case please

1) fix this PR's message
2) Implement Avinash's suggestion to remove the prepreamble from the disassembler tester though.

**Comment by edwardcwang** [comment]
One thought on the pre-preamble in RX testing: maybe create a separate test that includes pre-preamble and checks that it does nothing wrong

**Comment by jerryduan07** [comment]
> I see. In that case please
> 
> 1. fix this PR's message
> 2. Implement Avinash's suggestion to remove the prepreamble from the disassembler tester though.

Ah yeah, that is a typo
I'll remove pre-preamble in RX tests

File: src/test/scala/PacketAssembler_test/PacketDisassemblerTester.scala [hunk 1]
```diff
@@ -74,18 +74,29 @@ class PacketDisAssemblerTest(c: PacketDisAssembler) extends PeekPokeTester(c) {
     }
     poke(c.io.in.data.valid, false.B)
     step(10)
+    poke(c.io.in.data.valid, true.B)
+
+    for (j <- 0 to 7) {
```
Comment (edwardcwang) [inline src/test/scala/PacketAssembler_test/PacketDisassemblerTester.scala]: Maybe make the pre-preamble an optional part of the tester and you can test both?

Comment (jerryduan07) [reply src/test/scala/PacketAssembler_test/PacketDisassemblerTester.scala]: I can test pre-preamble in half of the 100 random packets

Comment (edwardcwang) [reply src/test/scala/PacketAssembler_test/PacketDisassemblerTester.scala]: Not sure I followed what you mean

Comment (jerryduan07) [reply src/test/scala/PacketAssembler_test/PacketDisassemblerTester.scala]: in packet disassembler test there is a for loop that generates 100 random packets and passes them through the module - i can test pre-preamble in 50 packets and not test it in the other 50

Comment (edwardcwang) [reply src/test/scala/PacketAssembler_test/PacketDisassemblerTester.scala]: Is that currently being done?

Comment (jerryduan07) [reply src/test/scala/PacketAssembler_test/PacketDisassemblerTester.scala]: it is now

**Review by avi-jois — APPROVED** [review]

**Review by avi-jois — CHANGES_REQUESTED** [review]
Can you make the prepreamble optional, set by an input? It's nice to have the configuration options where possible

**Review by avi-jois — APPROVED** [review]

**Pull request merged**